### PR TITLE
fix apiserver tls overwrite bug

### PR DESCRIPTION
**Pull request by adohe-zz**
Please refer #19309 for more detail.

### CODE REVIEW

**Comment by thockin** [comment]
cc @roberthbailey

File: pkg/genericapiserver/genericapiserver.go [hunk 1]
```diff
@@ -590,10 +591,12 @@ func (s *GenericAPIServer) Run(options *ServerRunOptions) {
 			alternateDNS := []string{"kubernetes.default.svc", "kubernetes.default", "kubernetes"}
 			// It would be nice to set a fqdn subject alt name, but only the kubelets know, the apiserver is clueless
 			// alternateDNS = append(alternateDNS, "kubernetes.default.svc.CLUSTER.DNS.NAME")
-			if err := util.GenerateSelfSignedCert(s.ClusterIP.String(), options.TLSCertFile, options.TLSPrivateKeyFile, alternateIPs, alternateDNS); err != nil {
-				glog.Errorf("Unable to generate self signed cert: %v", err)
-			} else {
-				glog.Infof("Using self-signed cert (%options, %options)", options.TLSCertFile, options.TLSPrivateKeyFile)
+			if !fileExists(options.TLSCertFile) || !fileExists(options.TLSPrivateKeyFile) {
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: I suggest:

* If only one of these files is present, or readable, do a fatal error-- that's probably a typo or something. The user should give us both or neither. (And what about the CA file? I guess we require that to be stuffed in with the cert?)

File: pkg/genericapiserver/genericapiserver.go [hunk 2]
```diff
@@ -627,6 +630,11 @@ func (s *GenericAPIServer) Run(options *ServerRunOptions) {
 	glog.Fatal(http.ListenAndServe())
 }
 
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return !os.IsNotExist(err)
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: This may be overly specific, e.g., what if the file is present but we don't have permission?

Comment (AdoHe) [inline pkg/genericapiserver/genericapiserver.go]: @lavalamp actually I just want to make sure this file exists here, if we get a permission error, of course the file exists. Guess you want to make sure this file exists and also readable(we have permission), I will update this to make it more robust and generic.

**Comment by AdoHe** [comment]
@lavalamp I just update this fix, now the basic idea is if both of `apiserver.crt` and `apiserver.key` exist and accessible, we will not generate self-signed files, otherwise we will try to generate self-signed certificates. What do you think about this?

File: pkg/genericapiserver/genericapiserver.go [hunk 3]
```diff
@@ -627,6 +630,25 @@ func (s *GenericAPIServer) Run(options *ServerRunOptions) {
 	glog.Fatal(http.ListenAndServe())
 }
 
+// If the file represented by path exists and
+// accessable, return true otherwise return false.
+func canReadFile(path string) bool {
+	if _, err := os.OpenFile(path, os.O_RDWR, 0666); err != nil {
+		return false
+	}
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: You need to close the file if you open it.

File: pkg/genericapiserver/genericapiserver.go [hunk 4]
```diff
@@ -627,6 +630,25 @@ func (s *GenericAPIServer) Run(options *ServerRunOptions) {
 	glog.Fatal(http.ListenAndServe())
 }
 
+// If the file represented by path exists and
+// accessable, return true otherwise return false.
+func canReadFile(path string) bool {
+	if _, err := os.OpenFile(path, os.O_RDWR, 0666); err != nil {
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Consider just using os.Open, I think it has defaults you want. (And you probably don't want O_RDWR anyway?)

Comment (AdoHe) [inline pkg/genericapiserver/genericapiserver.go]: I made a stupid mistake, sorry about that. I just update this, besides the certFile should be concatenation of apiserver's certificate and the CA's certificate, what about the CA file?

**Comment by AdoHe** [comment]
@lavalamp I just revisit this PR and noticed that this comment above `GenerateSelfSignedCert` method:
```
If the certificate or key files already exist, they will be overwritten.
```
So the default behavior of certs generation will overwrite the existing files, I don't know whether we should apply this PR. What do you think about?

**Comment by AdoHe** [comment]
@lavalamp kindly ping :)

**Comment by roberthbailey** [comment]
Adding the ok-to-merge label since the bot seems to be mis-behaving on this cl. @eparis

**Comment by lavalamp** [comment]
@AdoHe Sorry, I'm probably going to sit on this until after 1.2!

**Comment by AdoHe** [comment]
@lavalamp ok, we can leave this to next-candidate.

**Comment by lavalamp** [comment]
thanks for understanding :)

File: pkg/genericapiserver/genericapiserver.go [hunk 5]
```diff
@@ -680,6 +683,28 @@ func (s *GenericAPIServer) Run(options *ServerRunOptions) {
...
+		return false
+	}
+
+	defer f.Close()
+
+	return true
+}
+
+func shouldGenSelfSignedCerts(certPath, keyPath string) bool {
+	if canReadFile(certPath) && canReadFile(keyPath) {
```
Comment (lavalamp) [inline pkg/genericapiserver/genericapiserver.go]: Should be ||?

**Comment by lavalamp** [comment]
(I'm paying attention to PRs again-- ping me when ready!)

**Comment by AdoHe** [comment]
@lavalamp I revisit this PR and notice this comment above `GenerateSelfSignedCert ` method:
```
If the certificate or key files already exist, they will be overwritten.
```
For some consideration, the default behavior of `GenerateSelfSignedCert` will overwrite the existed certificate or key files, if we apply this PR, we will change the default behavior, I just want to make sure we did want to change this.

**Comment by AdoHe** [comment]
@lavalamp kindly ping :) also I am wondering whether we should make the same thing to kubelet.

**Comment by lavalamp** [comment]
Yes I think the point of this is to not run GenerateSelfSignedCert when it might overwrite something.

**Comment by lavalamp** [comment]
Yeah, if kubelet has a similar behavior, then probably?

**Comment by AdoHe** [comment]
Probably we should apply the same thing to kubelet, maybe we can leave this to other PR.

**Pull request merged**